crates/ecs/src/lib.rs

```diff
@@ -806,4 +1138,391 @@ mod tests {
+        world.store_entity_in_archetype(e3.id, 1);
+
+        // insert some components...
+        world.add_component::<u64>(e1.id, 1);
+        world.add_component::<u32>(e1.id, 2);
+
+        world.add_component::<u64>(e2.id, 3);
+        world.add_component::<u32>(e2.id, 4);
+
+        world.add_component::<u32>(e3.id, 6);
+
+        let mut result: Vec<u32> = vec![];
+
+        let mut borrowed = <Read<u32> as SystemParameter>::borrow(
+            &world,
+            &[<Read<u32> as SystemParameter>::signature()],
+        );
+        unsafe {
+            while let Some(parameter) =
+                <Read<u32> as SystemParameter>::fetch_parameter(&mut borrowed)
+            {
+                if let Some(parameter) = parameter {
+                    result.push(*parameter);
+                }
+            }
+        }
+
+        println!("{:?}", result);
+
+        assert_eq!(result, vec![2, 4, 6])
+    }
+
+    #[test]
+    #[should_panic]
+    fn borrowing_component_vec_twice_from_archetype_causes_panic() {
+        let mut archetype = Archetype::default();
+        archetype.add_component_vec::<u32>();
+
+        let borrow_1 = archetype.borrow_component_vec_mut::<u32>();
+        let borrow_2 = archetype.borrow_component_vec_mut::<u32>();
+
+        // Drop after both have been borrowed to make sure they both live this long.
+        drop(borrow_1);
+        drop(borrow_2);
+    }
+
+    #[test]
+    fn borrowing_component_vec_after_reference_has_been_dropped_does_not_cause_panic() {
+        let mut archetype = Archetype::default();
+        archetype.add_component_vec::<u32>();
+
+        let borrow_1 = archetype.borrow_component_vec_mut::<u32>();
+        drop(borrow_1);
+
+        let borrow_2 = archetype.borrow_component_vec_mut::<u32>();
+        drop(borrow_2);
+    }
+
+    #[test]
+    fn borrowing_two_different_component_vecs_from_archetype_does_not_cause_panic() {
+        let mut archetype = Archetype::default();
+        archetype.add_component_vec::<u32>();
+        archetype.add_component_vec::<u64>();
+
+        let a = archetype.borrow_component_vec_mut::<u32>();
+        let b = archetype.borrow_component_vec_mut::<u64>();
+
+        // Drop after both have been borrowed to make sure they both live this long.
+        drop(a);
+        drop(b);
+    }
+
+    #[test]
+    fn removing_entity_swaps_position_of_last_added_entitys_components() {
+        // Arrange
+        let mut archetype = Archetype::default();
+
+        archetype.store_entity(10);
+        archetype.store_entity(20);
+        archetype.store_entity(30);
+
+        archetype.add_component_vec::<u32>();
+        archetype.add_component_vec::<f32>();
+
+        archetype.add_component::<u32>(10, 1);
+        archetype.add_component::<u32>(20, 2);
+        archetype.add_component::<u32>(30, 3);
+
+        archetype.add_component::<f32>(10, 1.0);
+        archetype.add_component::<f32>(20, 2.0);
+        archetype.add_component::<f32>(30, 3.0);
```

Comment:
nit:
I think I preferred the way you handled it in other tests, where you named the entity indices `entity0_index`, `entity1_index` and so on... instead of using the indices directly like this (i.e. magic numbers). Using named variables makes it a lot clearer that this `10` is the same thing as this other `10`.